specs/codecs/wac/spec.md

```diff
@@ -0,0 +1,192 @@
+| 9 | VALUE_NULL |
+| 10 | VALUE_TRUE |
+| 11 | VALUE_FALSE |
+
+#### TYPE_LINK
+
+```
+| 0 | CID |
+```
+
+Note: Simple-DAG put a VARINT_LENGTH before the CID indicating how long it was.
+
+CIDs are self-delimiting so this didn't seem necessary.
+
+#### TYPE_INTEGER
+
+```
+| 1 | VARINT |
+```
+
+#### TYPE_SIGNED_INTEGER
+
+```
+| 2 | VARINT |
+```
+
+#### TYPE_FLOAT
+
+```
+| 3 | MATISSA_LENGTH | VARINT
+```
+
+TODO: Floats (and negative floats) need definition here.
+Making implementations actually bidirectional with respect to the IPLD Data Model seems difficult here.
+
+#### TYPE_NEGATIVE_FLOAT
+
+```
+| 4 | MATISSA_LENGTH | VARINT
+```
+
+#### TYPE_STRING
+
+```
+| 5 | VARINT_LENGTH | STRING
+```
+
+Note: This is essentially the same as Binary, but with a different token flag.
+
+As with the IPLD Data Model itself Strings are basically just Bytes with something to tag them as "string-like" but where UTF-8 is encouraged. See [Strings].
+
+#### TYPE_BINARY
+
+```
+| 6 | VARINT_LENGTH | BINARY
+```
+
+#### TYPE_MAP
+
+```
+| 7 | VARINT_NUM_PAIRS | PAIRS
+```
+
+`PAIRS` contains alternating keys then values concatenated. i.e. ` KEY1 | VALUE1 | KEY2 | VALUE2 ...`
+
+This codec does not have any form of canonical map sorting as that would make it ill-fitted to the IPLD Data Model.
+
+As in the IPLD Data Model map keys must be of type String, however as described in the String section the only distinction between Strings and Bytes are identifier hints.
+
+Note: Simple-DAG went with `KEYS_VARINT_LENGTH | VALUES_VARINT_LENGTH | KEYS | VALUES |`. Both seem doable,
+this approach seemed to make writing encoder/decoders really simple. However, adding in more length prefixes
+makes creating faster "zero copy" decoders very nice. It seems to mostly be a tradeoff for which side has to have bigger buffers, the encoder or the decoder.
+
+Note: We could assert that keys are just `VARINT_LENGTH | STRING` and remove the String token since it's always a string.
+It's some added complexity and really stops us from putting anything other than Strings in map keys, but that may not be too bad.
+
+#### TYPE_LIST
+
+```
+| 8 | VARINT_NUM_ELEMENTS | VALUES |
+```
+
+`VALUES` contains every value concatenated.
+
+Note: Simple-DAG went with `VARINT_LENGTH` (the size of the VALUES binary section) instead of `VARINT_NUM_ELEMENTS`
+and in the `VALUES` section had every value proceeded by the length of the value.
+
+Both seem doable, this approach seemed to make writing a decoder really simple. However, adding in more length prefixes makes creating faster "zero copy" decoders very nice. It seems to mostly be a tradeoff for which side has to have bigger buffers, the encoder or the decoder.
+
+#### TYPE_NULL
+
+```
+| 9 |
+```
+
+#### TYPE_TRUE
+
+```
+| 10 |
+```
+
+#### TYPE_FALSE
+
+```
+| 11 |
+```
+
+### Strings
+
+As with the IPLD Data Model itself Strings are basically just Bytes with something to tag them as "string-like" but where UTF-8 is encouraged.
```

Comment:
I still consider it a mistake that the IPLD Data Model defines strings as arbitrary bytes. There currently is no valid use of it ("valid" in the sense of that you cannot create valid data with existing specified DAG codecs). To my knowledge, strings as arbitrary bytes causes more harm then good (here are two examples of PL related stacks where things were fixed in retrospective due to this: [Fixing an interop issue in libp2p gossipsub](https://developer.actyx.com/blog/2020/07/27/libp2p-gossipsub), [FIP-0027](https://github.com/filecoin-project/FIPs/blob/2a3294118ca93968a3e531699820a73ca657bd78/FIPS/fip-0027.md)).

Also newer systems based on IPLD like Edelweiss, define [Strings as a list of characters](https://github.com/ipld/edelweiss/blob/d34b6b5bb97b4c8118567919248164e173eed758/doc/type-system.md#strings).

Systems outside of IPLD that have similar challenges, like the WebAssembly Interface Types, where interoperability between different programming languages is the goal, came to the conclusion that [Strings should be a list of Unicode scalar values](https://github.com/WebAssembly/interface-types/issues/135).

Recently the topic was again brought up by @Stebalien at https://github.com/ipld/ipld/issues/203.

Having talked with people about this topic a lot in the last few years, I haven't found people that thought defining strings the same way as bytes is a good idea. Hence it might be time to revisit this decision.

I bring this issue up as without WAC, I didn't really mind what the IPLD spec said about strings, as practically they were Unicode. With having WAC this would change. So I would be in favour if WAC would define Strings as list of Unicode characters.

Reply:
There are two issues with Strings == Unicode that I think we'd have to deal with if we wanted to make a change here.
1. Is there existing data that would break here, e.g. dag-cbor data where there are non-unicode strings?
2. Maps are currently defined to have String keys which IIUC it's quite problematic to assume Unicode-only keys here. Would this change to support non-String keys?

For example see #227 which is a spec for a fairly reasonable existing format called Bencode which is used by BitTorrent. A straight up change of Strings are now Unicode would make that format super miserable to make a codec for.

Reply:
The stand-out example is addresses in Filecoin used as map keys. Thankfully that got changed in a (very) recent upgrade so they're now strings (can't find the issue but this was recognised as a mistake post-launch, but just not super high priority to "fix"). There exists Filecoin chain data that we _can't_ read in JavaScript because it sticks `[]byte` into a CBOR `string` field. I think maybe Rust may be in the same boat? Maybe FVM forced this fix. I've been considering hacks in JS to get backdoor access to the bytes but it's a real mess.

Go is comfortable with `string([]byte(string(...))`, but nobody else is.

I'm mostly with @vmx on this. Although we do have the ability to hand-wave a little bit on this, suggest that Unicode is how it is supposed to work, but that if an implementation chooses to use byte arrays without a validation step then it becomes caveat emptor that your data may not be readable by other implementations (though hand-waving is kind of gross if we don't have to do it—like we do with codecs with existing data).

Reply:
From: https://github.com/ipld/ipld/pull/226#discussion_r909194368

> I think https://github.com/ipld/ipld/issues/203 is inevitable at some point, it's only the Go-exclusive folks that are torn on this

Maybe although IMO #203 isn't feasible unless something is done about maps. Restricting map keys to be Strings while also insisting the Strings must be Unicode makes it impossible to have map keys that are non-Unicode.

As I mentioned above, how does a codec like #227 get implemented sanely? My understanding is there are also folks who want bytes or integer map keys that have been hobbling by with just shoving arbitrary stuff into bytes.

---

For others interested in the "what do we do about Strings" debate. I'd recommend reading https://github.com/ipld/ipld/blob/257834ab78bca17506e5ef0c4a7f00a43a06b0c0/notebook/exploration-reports/2020.10-data-model-maps-keys.md and the links from that issue. @warpfork and @vmx lay out a lot of the space and tradeoffs here for us to figure out what to do.

I have some opinions on this (e.g. the "IMO" section above), but if we're planning on making changes to the data model I'm certainly happy to reflect them in WAC since it's just trying to be a high fidelity representation that can support almost all of what the data model can.

Reply:
> 1. Is there existing data that would break here, e.g. dag-cbor data where there are non-unicode strings?

Even if, to me that's not really a reason not to change it. If it exists, it's invalid CBOR. I don't think we can (nor should) support any invalid CBOR data that is out there. It was produced by buggy implementation, not due to being within the bounds of specs.

Reply:
About non-string map keys. AFAICT this is one of the biggest reasons why currently Strings are defined the same way as bytes. I personally prefer the workaround of having lists of two-tuples `(key, value)`, where then key can be of any type. So I don't have experience how practical that is in real world scenarios.

Another potential change to the data model could be (I'm not sure I like it), is to make map keys of type bytes, which reflects what they currently are.

I prefer the "tuple list" approach as I would expect that "strings as keys" is the majority of the use cases and I'm a fan of making the usual usage simple, while making more advanced things possible. If map keys would be bytes, you probably end up with all implementations checking for "is this a string, if not, do xyz", which sounds pretty painful. Instead I'd try to make those tuple lists a 1.5 class citizen. This would then even enable fixed, application specific ordering of maps. More on that idea can be found at https://github.com/ipld/ipld/blob/master/notebook/exploration-reports/2020.10-schema-listpair-extension.md.

Reply:
Just to jump on the bytes vs unicdoe thing. I'd like to use Maps to represent B trees down the line which rely on byte-ordering to create search indexes. Having a hard requirement of needing strings makes it a lot harder to efficiently represent index keys that take advantage of this byte ordering.

Does this mean there's a need here for something "like" maps but aligns more with HashMaps in other languages where a key could be arbitrary data? 🤷

Reply:
> I'd like to use Maps to represent B trees down the line which rely on byte-ordering to create search indexes.

Wouldn't tuple lists work for that quite well? You could have bytes and arbitrary ordering.

Reply:
> Wouldn't tuple lists work for that quite well? You could have bytes and arbitrary ordering.

That is technically possible but is not desired.

Reply:
> That is technically possible but is not desired.

What would be needed to make it desired? Would it help if it would be somehow implemented directly in IPLD libraries? To me (I might be the only one, but I'd like to change that :) the tuple list pattern is the best solution to the "arbitrary IPLD as map keys" and "custom sort order" problems.

Reply:
Could you elaborate on what the API would look like for a tuple list?

One of the main things I like the existing Map API is that it's easy to reason about with existing ADLs like HAMT. Having more moving bits and new APIs feels like it'll be harder to integrate with existing things that are willing to work with a regular Map.

Reply:
> Could you elaborate on what the API would look like for a tuple list?

I sadly can't I know too little about Go, go-ipld-prime and ADLs.

Reply:
@RangerMauve since you're talking about an ADL you just need to make sure you provide a sane mapping between encoded bytes and the external API, so it really shouldn't matter here whether you encode as Bytes or Strings. If you need arbitrary bytes then you encode as Bytes but maybe those get presented to the user through the ADL as Strings.

We do something like this with the HAMT and I'd expect to see a similar thing when we end up doing sorted maps: https://ipld.io/specs/advanced-data-layouts/hamt/spec/#schema - note how `BucketEntry` uses `key Bytes`, but when that gets all translated through the ADL code we end up with Strings (for the most part - we also retain the option of having Bytes keys through the API which isn't entirely precluded but is a touch controversial), see:

* https://github.com/rvagg/js-ipld-hashmap/blob/master/interface.ts
* https://github.com/ipld/go-ipld-adl-hamt/blob/9004dbd839e0e73ec52d9b1b47ddbd0627d74b92/node.go#L83-L87

So, thanks to ADLs, or "lenses" in general, we get to do the right thing at the codec layer, but mess around with it however we like above it to transform it into what the end user wants.

Same goes with schema "typed" interfaces, using "tuple" representation maps (since we're talking tuples here), which are encoded by the codec as a List but are presented to the user as a Map, or even a Struct (the Filecoin chain makes use of this extensively, not encoding _any_ Map to bytes even though they're presented to users as Structs that kind of look like they should be maps, they're just arrays of values in a certain order).